x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/MlConfigVersion.java

```diff
@@ -0,0 +1,389 @@
+                            + fieldName
+                            + "] have the same version number ["
+                            + version.id
+                            + "]. Each MlConfigVersion should have a different version number";
+                }
+            }
+        }
+
+        return Collections.unmodifiableNavigableMap(builder);
+    }
+
+    private static final NavigableMap<Integer, MlConfigVersion> VERSION_IDS = getAllVersionIds(MlConfigVersion.class);
+
+    static Collection<MlConfigVersion> getAllVersions() {
+        return VERSION_IDS.values();
+    }
+
+    public static MlConfigVersion readVersion(StreamInput in) throws IOException {
+        return fromId(in.readVInt());
+    }
+
+    public static MlConfigVersion fromId(int id) {
+        MlConfigVersion known = VERSION_IDS.get(id);
+        if (known != null) {
+            return known;
+        }
+        // this is a version we don't otherwise know about - just create a placeholder
+        return new MlConfigVersion(id, id >= MINIMUM_COMPATIBLE.id());
+    }
+
+    public static MlConfigVersion fromId(int id, boolean isLegacy) {
+        MlConfigVersion known = VERSION_IDS.get(id);
+        if (known != null) {
+            return known;
+        }
+        // this is a version we don't otherwise know about - just create a placeholder
+        return new MlConfigVersion(id, isLegacy);
+    }
+
+    public static MlConfigVersion fromNode(DiscoveryNode node) {
+        return fromString(node.getAttributes().get(ML_CONFIG_VERSION_NODE_ATTR));
+    }
+
+    public static void writeVersion(MlConfigVersion version, StreamOutput out) throws IOException {
+        out.writeVInt(version.id);
+    }
+
+    /**
+     * Returns the minimum version of {@code version1} and {@code version2}
+     * Legacy, semantic style representations are treated as being before all new, single number representations
+     */
+    public static MlConfigVersion min(MlConfigVersion version1, MlConfigVersion version2) {
+        if (version1.isLegacy != version2.isLegacy) {
+            return version1.isLegacy ? version1 : version2;
+        }
+
+        return version1.id < version2.id ? version1 : version2;
+    }
+
+    /**
+     * Returns the maximum version of {@code version1} and {@code version2}
+     * New, single number representations are treated as being after all legacy, semantic style representations.
+     */
+    public static MlConfigVersion max(MlConfigVersion version1, MlConfigVersion version2) {
+        if (version1.isLegacy != version2.isLegacy) {
+            return version1.isLegacy ? version2 : version1;
+        }
+
+        return version1.id > version2.id ? version1 : version2;
+    }
+
+    /**
+     * Returns {@code true} if the specified version is compatible with this running version of Elasticsearch.
+     */
+    public static boolean isCompatible(MlConfigVersion version) {
+        return version.onOrAfter(MINIMUM_COMPATIBLE);
+    }
```

Comment:
This should be removed. It's not _necessarily_ the case that every ML config created prior to 7.0 is invalid now. The decision needs to be taken on a per-config basis by the code that deals with that config.